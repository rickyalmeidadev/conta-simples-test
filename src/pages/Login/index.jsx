import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { login, storeToken } from '../../services/services';

const Login = () => {
  const [formData, setFormData] = useState({
    cpf: '',
    email: '',
    password: '',
  });
  const [userError, setUserError] = useState(false);
  const [reqError, setReqError] = useState(false);

  const history = useHistory();

  const handleChange = (event) => {
    const { name, value } = event.target;

    setFormData({ ...formData, [name]: value });
  };

  const handleCpf = (event) => {
    let { value } = event.target;

    if (/[A-Za-z]/.test(value)) return;

    if (value.length >= 11) {
      value = value.replace(/(\d{3})\.(\d{3})\.(\d{3})-/, '$1.$2.$3');
      value = value.replace(/(\d{3})\.(\d{3})\.(\d{3})(\d+)/g, '$1.$2.$3-$4');
    } else if (value.length >= 7) {
      value = value.replace(/(\d{3})\.(\d{3})\./g, '$1.$2');
      value = value.replace(/(\d{3})\.(\d{3})(\d+)/g, '$1.$2.$3');
    } else if (value.length >= 3) {
      value = value.replace(/(\d{3})\./g, '$1');
      value = value.replace(/(\d{3})(\d+)/g, '$1.$2');
    }

    setFormData({ ...formData, cpf: value });
  };

  const handleSubmit = async (event) => {
    event.preventDefault();

    const { cpf, email, password } = formData;

    if (!cpf || !email || !password) {
      setUserError(true);
      return;
    }
    try {
      const response = await login(email, password);
      storeToken(response.data.accessToken);
      history.push('/');
    } catch (error) {
      setReqError(true);
    }
  };

  const { cpf, email, password } = formData;

  return (
    <form onSubmit={handleSubmit}>
      <fieldset>
        <legend>Entrar</legend>

        <label htmlFor="cpf">
          CPF
          <input
            type="text"
            id="cpf"
            name="cpf"
            maxLength="14"
            value={cpf}
            onChange={handleCpf}
            data-testid="cpf"
          />
        </label>

        <label htmlFor="email">
          E-mail
          <input
            type="email"
            id="email"
            name="email"
            value={email}
            onChange={handleChange}
            data-testid="email"
          />
        </label>

        <label htmlFor="password">
          Senha
          <input
            type="password"
            id="password"
            name="password"
            value={password}
            onChange={handleChange}
            data-testid="password"
          />
        </label>

        {userError && (
          <span data-testid="user-error">
            Por favor, insira todos os dados.
          </span>
        )}
        {reqError && <span data-testid="req-error">Dados inv√°lidos.</span>}

        <button type="submit" data-testid="submit">
          Entrar
        </button>
      </fieldset>
    </form>
  );
};

export default Login;
