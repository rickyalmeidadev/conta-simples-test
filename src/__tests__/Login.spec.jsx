import React from 'react';
import {
  render,
  fireEvent,
  waitForElement,
  wait,
} from '@testing-library/react';
import Login from '../pages/Login';

const mockedHistoryPush = jest.fn();

jest.mock('../services/services', () => ({
  login: jest.fn(() => ({ data: { accessToken: 'valid_token' } })),
  storeToken: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: mockedHistoryPush,
  }),
}));

describe('Login in the DOM', () => {
  it('renders', () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('has a form for login', () => {
    const { getByTestId } = render(<Login />);

    expect(getByTestId('cpf')).toBeDefined();
    expect(getByTestId('email')).toBeDefined();
    expect(getByTestId('password')).toBeDefined();
    expect(getByTestId('submit')).toBeDefined();
  });
});

describe('Login functionalities', () => {
  it('shows error messages for empty fields', async () => {
    const { getByTestId } = render(<Login />);

    fireEvent.click(getByTestId('submit'));

    const error = await waitForElement(() => getByTestId('user-error'));

    expect(error).toBeDefined();
  });

  it('should login successfully if valid credentials are provided', async () => {
    const { getByTestId } = render(<Login />);

    const cpfInputElement = getByTestId('cpf');
    const emailInputElement = getByTestId('email');
    const passwordInputElement = getByTestId('password');
    const submitButtonElement = getByTestId('submit');

    fireEvent.change(cpfInputElement, { target: { value: '12345678910' } });
    fireEvent.change(emailInputElement, {
      target: { value: 'valid_email@mail.com' },
    });
    fireEvent.change(passwordInputElement, {
      target: { value: 'valid_password' },
    });

    fireEvent.click(submitButtonElement);

    await wait(() => {
      expect(mockedHistoryPush).toHaveBeenCalledWith('/');
    });
  });
});
