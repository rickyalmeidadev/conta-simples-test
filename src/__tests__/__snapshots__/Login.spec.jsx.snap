// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login in the DOM renders 1`] = `
<DocumentFragment>
  <div
    class="wrapper wrapper--row login-bg"
  >
    <main
      class="container container--center"
    >
      <form
        class="login"
      >
        <figure
          class="login__brand"
        >
          <img
            alt="Conta Simples"
            src="logo-conta-simples.svg"
          />
        </figure>
        <fieldset
          class="login__box"
        >
          <legend
            class="login__title"
          >
            Preencha seus dados
          </legend>
          <label
            class="label"
            for="cpf"
          >
            CPF
            <input
              class="input"
              data-testid="cpf"
              id="cpf"
              maxlength="14"
              name="cpf"
              type="text"
              value=""
            />
          </label>
          <label
            class="label"
            for="email"
          >
            E-mail
            <input
              class="input"
              data-testid="email"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </label>
          <label
            class="label"
            for="password"
          >
            Senha
            <input
              class="input"
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </label>
          <button
            class="btn"
            data-testid="submit"
            type="submit"
          >
            Entrar
          </button>
        </fieldset>
      </form>
    </main>
  </div>
</DocumentFragment>
`;
